docs/design/tanka.md

```diff
@@ -0,0 +1,223 @@
+| `baseDir`        | `main.jsonnet`     | Environment specific code can be put here. The special `main.jsonnet` is the entry point for the evaluation                           |
+
+To resolve the `JPATH`, tanka first traverses the directory tree *upwards*, to
+find a `jsonnetfile.json`, which marks the `rootDir`. Reaching `/` without a
+match will result in an error.
+
+Same applies for the `baseDir`, the tree is traversed *upwards* for a
+`main.jsonnet`.
+
+The final `JPATH` looks like the following:
+```
+<baseDir>
+<rootDir>/lib
+<rootDir>/vendor
+```
+
+
+### Directory structure
+In a simple setup, it is fair to have the same `rootDir` and `baseDir`:
+```
+.
+├── jsonnetfile.json
+├── lib/
+├── main.jsonnet
+└── vendor/
+```
+
+However, to use [Environments](#environments), the `baseDir` could be moved
+into subdirectories:
+```
+.
+├── environments
+│   ├── dev
+│   │   └── main.jsonnet
+│   └── prod
+│       └── main.jsonnet
+├── jsonnetfile.json
+├── lib/
+└── vendor/
+```
+
+While latter structure is the one suggested by `tk init`, it is perfectly fine to
+use another if it fits the use-case better. The folder does not need to be named
+`environments`, either.
+
+## Edge Cases
+During development of `jsonnet` libraries, e.g. for applications like `mysql`,
+it is impossible to think of every edge-case in before.
+
+But thanks to the power of `jsonnet`, this is not a problem. Imagine the output
+of the library being the following:
+```jsonnet
+local out = {
+  apiVersion: "v1",
+  kind: "namespace",
+  metadata: {
+    name: "production"
+  }
+};
+```
+
+When you wanted to label the namespace, but the library did not provide
+functions for it, you could use deep-merging:
+```jsonnet
+out + {
+  metadata+: {
+    labels: {
+      foo: "bar"
+    }
+  }
+}
+```
+
+Note the special `+:` to enable deep merging.
+
+This would result in the second dict being recursively merged on top of the first one:
+```jsonnet
+{
+  apiVersion: "v1",
+  kind: "namespace",
+  metadata: {
+    name: "production",
+    labels: {
+      foo: "bar"
+    }
+  }
+}
+```
+
+## Environments
+The only core concept of tanka is an `Environment`. It describes a single
+context that can be configured. Such a context might be `dev` and `prod`
+environments, Blue/Green deployments or the same application available
+in multiple regional zones / datacenters.
+
+An environment does not need to be created, it rather just exists as soon as a
+`main.jsonnet` is found somewhere in the tree of a `rootDir`.
+
+However, an environment may receive additional configuration, by adding a file
+called `spec.json` alongside the `main.jsonnet`:
+```json
+{
+  "apiVersion": "tanka.dev/v1alpha1",
+  "kind": "Environment",
+  "metadata": {
+    "name": "auto",
+    "labels": {}
+  },
+  "spec": {
+    "apiServer": "https://localhost:6443",
+    "namespace": "default"
+  }
+}
+```
```

Comment:
Note, we have extended our spec.json to include "team" i.e. the team that manages this environment, and "status", e.g. whether it is inactive, or a dev environment, etc. I don't think these are special cases that need enshrining in this doc, but perhaps mention that it is possible, or maybe even add a node within the file for custom descriptors.

The use case would be "find all prod environments" or "find all environments managed by team X".

I presume you are intentionally diverging from the current ksonnet format of the spec.json file (it wouldn't be hard to fix them all to match this format).

Reply:
The current format is still supported, but `deprecated`. 
The team parameter is only descriptive, so it can be put into `metadata.labels`